### PR TITLE
Implement zip() function for dataset

### LINKED ISSUE #18580: Zip two datasets together
### Describe your feature request
We have two dataset in the ObjectStore, 
dataset1 = Dataset(num_bcloks=10, num_rows=100, schema={id: Int64, feat_1:double,feat_2:Int64, feat_3:Double)
dataset2 = Dataset(num_bcloks=10, num_rows=100, schema={id: Int64, score_1:double,score_2:Int64, score_3:Int64)
Id is a one-to-one correspondence in the two dataset，We want a new dataset with schema={id:Int64, feat_1:double,score_1:double}，Is there any related,API that can achieve this funcation。
image

For questions, post on our Discussion page instead: https://discuss.ray.io/

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

Closes https://github.com/ray-project/ray/issues/18580

### CODE REVIEW

File: python/ray/data/dataset.py [hunk 1]
```diff
@@ -655,6 +655,55 @@ def sort(self,
...
+        (e.g., one was produced from a ``.map()`` of another).
+
+        Time complexity: O(dataset size / parallelism)
+
+        Args:
+            other: The dataset to zip with on the right hand side.
+
+        Examples:
+            >>> ds1 = ray.data.range(5)
+            >>> ds2 = ray.data.range(5).map(lambda x: x + 1)
```
Comment (scv119) [inline python/ray/data/dataset.py:671]: incorrect comments?

File: python/ray/data/dataset.py [hunk 2]
```diff
@@ -655,6 +655,55 @@ def sort(self,
...
+
+        Returns:
+            A Dataset with (k, v) pairs (or concatenated Arrow schema) where k
+            comes from the first dataset and v comes from the second.
+        """
+
+        blocks1 = self.get_blocks()
+        blocks2 = other.get_blocks()
+
+        if len(blocks1) != len(blocks2):
```
Comment (scv119) [inline python/ray/data/dataset.py:683]: should we add a TODO to support zip two dataset with different number of blocks, but the number rows are identical?

File: python/ray/data/impl/arrow_block.py [hunk 3]
```diff
@@ -200,6 +200,30 @@ def num_rows(self) -> int:
     def size_bytes(self) -> int:
         return self._table.nbytes
 
+    def zip(self, other: "Block[T]") -> "Block[T]":
```
Comment (scv119) [inline python/ray/data/impl/arrow_block.py:203]: worth document the behavior of zipping two blocks with duplicate columns: undefined/overwrites/etc?

**Comment by scv119** [comment]
mostly nits.

**Comment by ericl** [comment]
Comments addressed.

**Review by scv119 — APPROVED** [review]

**Pull request merged**